### PR TITLE
Added NMI feature selector

### PR DESCRIPTION
**Pull request by DmitriiZorin**
# NMI feature selectior 
NMI feature selectior is an algorithm for feature selection based on mutual information score and its normalization (paper https://ieeexplore.ieee.org/document/10589754). 
Implemented with usage of pyitlib for entropy calculation

# Usage example
Here is a usage example for feature selection in dataset from file dataset.csv with NMIFeatureSelector with threshhold 0.7.
```
df = pandas.read_csv('dataset.csv')
selector = NMIFeatureSelector(threshhold=0.7)
selector.fit(dataset, 'target feature name')
print(selector.transform())
```

### CODE REVIEW

**Comment by jrzkaminski** [comment]
Hello! Thanks for the PR.

Here is a general list of what has to be added:

### What's Missing:
1. Docstrings should be [Google-stylized](https://google.github.io/styleguide/pyguide.html), so please, add arguments to the docstrings and format them.
2. Tests. Please, add unit tests using pytest library for the module.
3. Documentation. The documentation should include a brief explanation of how the algorithm works and a short usage example. A placeholder documentation for this module already exists [here](https://github.com/Anaxagor/applyBN/blob/main/applybn/feature_selection/bn_feature_selector.py). It fully supports markdown, so should not be a problem.
4. In-line comments. Please, add them so it is easier to understand what's going on in the code.
5. [Typing](https://docs.python.org/3/library/typing.html). Please, add typing hints to arguments of methods and class.

File: applybn/feature_selection/bn_nmi_feature_selector.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    for e1, e2 in zip(x, y):
+      if (e1,e2) not in discr:
+        discr[(e1,e2)] = 0
+      discr[(e1,e2)] += 1
+    lenx = len(self._discreticise(x)) + 1
+    leny = len(self._discreticise(y)) + 1
+    res = [0] * (lenx * leny)
+    for k, v in discr.items():
+      res[(k[0] * leny + k[1])] = v
+    return res
```
Comment (jrzkaminski) [inline applybn/feature_selection/bn_nmi_feature_selector.py:34]: A decent effort to do things on your own, but why not just use numpy? I am unsure, why do you need an y here if you count occurrences of one variable, but you get the idea. So please, use numpy to omit these inefficient loops.

Here is little code snippet of how it could be done with numpy. You also can look into numpy.unique.
```python
import numpy as np

@staticmethod
def _discreticise(x, y=None):
    """
    Discretize one or two arrays using NumPy for efficiency.
    If y is provided, returns the flattened joint histogram (2D).
    Otherwise, returns a one-dimensional count of occurrences in x.
    """
    arr_x = np.array(x, dtype=int)
    
    if y is None:
        # Single array discretization using bincount
        offset = arr_x.min()
        shifted = arr_x - offset  # Shift values to start at 0
        counts = np.bincount(shifted)
        return counts.tolist()
    else:
        # Two-array discretization using histogram2d
        arr_y = np.array(y, dtype=int)
        x_min, x_max = arr_x.min(), arr_x.max()
        y_min, y_max = arr_y.min(), arr_y.max()

        # bins=[...] ensures each integer has its own bin
        hist, x_edges, y_edges = np.histogram2d(arr_x, arr_y,
                                                bins=[x_max - x_min + 1,
                                                      y_max - y_min + 1],
                                                range=[[x_min, x_max + 1],
                                                       [y_min, y_max + 1]])
        return hist.flatten().astype(int).tolist()

# Example usage:
if __name__ == "__main__":
    x_vals = [0, 1, 1, 2, 2, 2, 3]
    y_vals = [1, 1, 2, 2, 3, 3, 3]

    # Single-array discretization
    single_result = discreticise(x_vals)
    print("Single-array result:", single_result)

    # Two-array (joint) discretization
    joint_result = discreticise(x_vals, y_vals)
    print("Joint-array result:", joint_result)
```


Event better, you could use native python tool Counter:

```python
from collections import Counter

def discreticise_single(x):
    counts = Counter(x)
    # Convert counts to a list if needed
    min_val, max_val = min(x), max(x)
    result = [counts[i] for i in range(min_val, max_val+1)]
    return result
``` 

What's faster: numpy or Counter? Here is the answer:
https://stackoverflow.com/questions/52969157/numpy-unique-vs-collections-counter-performance-question

**Review by jrzkaminski — CHANGES_REQUESTED** [review]

File: applybn/feature_selection/bn_nmi_feature_selector.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    return (Ha + Hb - Hab) / (2 * max(Ha, Hb))
+
+  def _prep_dataset(self, df, target):
+    '''
+    Prepare the dataset, discretize the fields, remove rows
+    Return a ready-to-use dataset
+    '''
+    clean_df = pd.DataFrame()
+    for col in df:
+      if df[col].dtype in ['int', 'int32','int64']:
```
Comment (jrzkaminski) [inline applybn/feature_selection/bn_nmi_feature_selector.py:60]: The code checks if `df[col].dtype` is in `['int','int32','int64']` or `['float','float64','double']`. It might be safer to use pandas API (e.g., `pd.api.types.is_integer_dtype` or `is_numeric_dtype`) to handle potential variations.

File: applybn/feature_selection/bn_nmi_feature_selector.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    self.verboose = verboose
+
+
+  def _discreticise(self, x, y=None):
+    '''
+    Prep data to feed into pyitlib drv.entropy
+    Calc happenings of each event in descreticised feature
+    '''
+    if y is None:
+      discr = [0] * (max(x) - min(x) + 1)
```
Comment (jrzkaminski) [inline applybn/feature_selection/bn_nmi_feature_selector.py:19]: Consider what happens if columns are empty, or if max(x) == min(x). This would create a list of length 1 in _discreticise, which may cause edge cases.

File: applybn/feature_selection/bn_nmi_feature_selector.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+        clean_df[col] = clean_df[col].cat.add_categories(self.bin_count + 1).fillna(self.bin_count + 1)
+      elif df[col].dtype in ['string', 'object']:
+        # pandas cut does not work with string features, this is done manually
+        # it is assumed that the user has removed the ID-like features, otherwise there will be a memory error later
+        df_map = { l:i for i,l in enumerate(set(df[col].to_list()))}
+        clean_df[col] = df[col].map(df_map)
+    return clean_df
+
+  def fit(self, df, target):
+    clean_df = self._prep_dataset(df, target)
```
Comment (jrzkaminski) [inline applybn/feature_selection/bn_nmi_feature_selector.py:76]: Depending on dataset size, repeated use of set(...) or enumerating large columns may be slow.
If large amounts of data are being processed, you might consider vectorization with NumPy or dedicated libs (like scikit-learn’s KBinsDiscretizer) for binning float columns.

File: applybn/feature_selection/bn_nmi_feature_selector.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+    if self.verboose:
+      print('Dataset is ready')
```
Comment (jrzkaminski) [inline applybn/feature_selection/bn_nmi_feature_selector.py:78]: Please, replace print statements with logger

File: applybn/feature_selection/bn_nmi_feature_selector.py [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+        df_map = { l:i for i,l in enumerate(df[col].unique())}
+        clean_df[col] = df[col].map(df_map)
+    return clean_df
+
+  def fit(self, X: pd.DataFrame, y: pd.Series = None):    
+    if y is None:
+      return self
+
+    df = X
+    df[y.name] = y
```
Comment (jrzkaminski) [inline applybn/feature_selection/bn_nmi_feature_selector.py:100]: Since X is a dataframe, it is also mutable. When a user passes his data into this function, the original data will be modified later on. Because variables data, X, and df will refer to the same DataFrame object in memory. This is an undesirable behavior in most cases.
TL;DR use pandas `.copy()` method.

File: applybn/feature_selection/bn_nmi_feature_selector.py [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+    self.entropy_map = dict()
+    pre_features = []
+    for col in clean_df:
+      if col == target:
+        continue
+      if (col, target) not in nmi_map.keys():
+        nmi_map[(col, target)] = self._normalized_mutual_information(clean_df[col], clean_df[target])
+      nmi = nmi_map[(col, target)]
+      if nmi > self.threshhold:
+        pre_features.append(col)
```
Comment (jrzkaminski) [inline applybn/feature_selection/bn_nmi_feature_selector.py:117]: Here, I assume, that various pandas methods could be used to avoid running pandas inside loop the same as in `_prepare_dataset` method.

File: applybn/feature_selection/bn_nmi_feature_selector.py [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+        if (col1, col2) not in nmi_map:
+          nmi_map[(col1, col2)] = self._normalized_mutual_information(clean_df[col1], clean_df[col2])
+
+        nmi_tg_1 = nmi_map[(col1, target)]
+        nmi_tg_2 = nmi_map[(col2, target)]
+        nmi_col_12 = nmi_map[(col1, col2)]
+
+        if (nmi_tg_1 > nmi_tg_2) and (nmi_col_12 > nmi_tg_2):
+          if col1 not in features:
+            features.append(col1)
```
Comment (jrzkaminski) [inline applybn/feature_selection/bn_nmi_feature_selector.py:147]: nested loops here also. Consider refactoring that with pandas methods to avoid nested loops.

**Review by jrzkaminski — CHANGES_REQUESTED** [review]
Hi! Thanks for the improvements.
Everything is good, however, some minor changes are required.

File: applybn/feature_selection/bn_nmi_feature_selector.py [hunk 9]
```diff
@@ -0,0 +1,166 @@
+from pyitlib import discrete_random_variable as drv
+from sklearn.feature_selection import SelectorMixin
+from sklearn.base import BaseEstimator
+import pandas as pd
+import numpy as np
+import logging
+import itertools
+
+
+class NMIFeatureSelectior(BaseEstimator, SelectorMixin):
```
Comment (jrzkaminski) [inline applybn/feature_selection/bn_nmi_feature_selector.py:10]: Typo in **Selector**
```suggestion
class NMIFeatureSelector(BaseEstimator, SelectorMixin):
```

File: applybn/feature_selection/bn_nmi_feature_selector.py [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+    
+    Args:
+        threshold (float): The minimum NMI to include a feature in the selection.
+        bin_count (int): Number of bins for float discretization.
+        verbose (bool): If True, logging intermediate information.
+    '''
+    self.threshhold = threshhold
+    self.bin_count = bin_count
+    self.bin_labels = [i for i in range(self.bin_count)]
+    self.verboose = verboose
```
Comment (jrzkaminski) [inline applybn/feature_selection/bn_nmi_feature_selector.py:24]:
```suggestion
    self.verbose = verbose
```

File: applybn/feature_selection/bn_nmi_feature_selector.py [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+            nmi_map[(col1, col.name)] = self._normalized_mutual_information(clean_df[col1], col)
+
+          nmi_tg_1 = nmi_map[(col1, target)]
+          nmi_tg_2 = nmi_map[(col.name, target)]
+          nmi_col_12 = nmi_map[(col1, col.name)]
+
+          if (nmi_tg_1 > nmi_tg_2) and (nmi_col_12 > nmi_tg_2):
+            if col1 not in features:
+              features.append(col1)
+        return col
```
Comment (jrzkaminski) [inline applybn/feature_selection/bn_nmi_feature_selector.py:151]: A function is declared inside of a loop.

**Review by jrzkaminski — APPROVED** [review]
